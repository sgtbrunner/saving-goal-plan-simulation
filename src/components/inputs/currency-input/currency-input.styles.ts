import styled from 'styled-components';
import CurrencyInput from 'react-currency-input-field';

export const IconContainer = styled.div`
  margin: 20px 14px 20px 18px;
  color: var(--blue-gray-100-color);
`;

export const Input = styled(CurrencyInput)`
  border: none;
  color: var(--blue-gray-600-color);
  font-family: var(--currency-font-family);
  font-size: var(--md-font-size);
  font-weight: var(--normal-font-weight);
  line-height: var(--md-line-height);
`;
