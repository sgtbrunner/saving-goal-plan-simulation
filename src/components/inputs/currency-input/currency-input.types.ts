export type CurrencyInputProps = {
  label: string;
  name?: string;
  symbol?: string;
  initialValue?: string | number;
};
