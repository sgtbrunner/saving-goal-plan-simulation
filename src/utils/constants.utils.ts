// OPERATIONAL CONSTANTS
export const FIRST_STEP = 1;
export const LOCALE = 'en';
export const INITIAL_GOAL_AMOUNT = 25000;
export const PREVIOUS_BUTTON_ID = 'previous-month';
export const NEXT_BUTTON_ID = 'next-month';
export const ARROW_RIGHT = 'ArrowRight';
export const ARROW_LEFT = 'ArrowLeft';

// GENERAL CONSTANTS
export const ID = 'id';
export const ORIGIN = 'Origin';
export const LEFT = 'left';
export const RIGHT = 'right';
export const WHITE = 'white';
export const DOLLAR_SIGN = '$';
