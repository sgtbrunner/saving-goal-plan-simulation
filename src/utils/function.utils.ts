export const clearInvalidCharacters = (value: string): string =>
  value.replace(/\D|^0+/g, '');

export const convertInputToNumber = (value: string): number => {
  if (value.includes('.')) {
    const [decimalPart, fractionalPart] = value.split('.');
    return Number(
      [
        clearInvalidCharacters(decimalPart),
        '.',
        clearInvalidCharacters(fractionalPart),
      ].join('')
    );
  }

  return Number(value.replace(/\D|^0+/g, ''));
};
