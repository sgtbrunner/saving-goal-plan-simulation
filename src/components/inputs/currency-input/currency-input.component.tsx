import React, { Fragment, useState } from 'react';

import { convertInputToNumber } from '../../../utils/function.utils';
import { CurrencyInputProps } from './currency-input.types';
import {
  CurrencyInputContainer,
  InputLabel,
  IconContainer,
  Input,
} from './currency-input.styles';

const CurrencyInput = ({
  label,
  name = 'amount',
  symbol = '$',
  initialValue = '25000',
}: CurrencyInputProps): JSX.Element => {
  const [amount, setAmount] = useState(initialValue);

  const handleChange = ({
    target,
  }: React.ChangeEvent<HTMLInputElement>): void => {
    const el = target as HTMLInputElement;
    setAmount(convertInputToNumber(el.value));
  };

  return (
    <Fragment>
      <InputLabel htmlFor={name}>{label}</InputLabel>
      <CurrencyInputContainer>
        <IconContainer>
          <span>{symbol}</span>
        </IconContainer>
        <Input
          id={name}
          name={name}
          allowNegativeValue={false}
          decimalSeparator="."
          groupSeparator=","
          onChange={handleChange}
          disableAbbreviations
          defaultValue={amount}
        />
      </CurrencyInputContainer>
    </Fragment>
  );
};

export default CurrencyInput;
