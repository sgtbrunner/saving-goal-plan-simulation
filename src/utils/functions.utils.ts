import { FIRST_STEP, LOCALE } from './constants.utils';
import { MonthPickerInputState } from './types.utils';

export const getDate = (customDate?: string | number): Date =>
  customDate ? new Date(customDate) : new Date();

export const monthFormatter = new Intl.DateTimeFormat(LOCALE, {
  month: 'long',
});

export const yearFormatter = new Intl.DateTimeFormat(LOCALE, {
  year: 'numeric',
});

export const getMonthPickerState = (
  step = FIRST_STEP,
  modifier = 0
): MonthPickerInputState => {
  const today = getDate();
  const targetMonthISO = today.setMonth(today.getMonth() + step + modifier, 1);

  return {
    step: step + modifier,
    year: yearFormatter.format(targetMonthISO),
    month: monthFormatter.format(targetMonthISO),
  };
};

export const getSignificantAndFractionalDigits = (
  value: string
): Array<string> => value.split('.');
