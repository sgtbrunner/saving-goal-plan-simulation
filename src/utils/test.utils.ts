import { INITIAL_GOAL_AMOUNT } from './constants.utils';
import { getDate, monthFormatter, yearFormatter } from './functions.utils';
import { SavingGoalState } from './types.utils';

// TAGS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const IMG = 'img';
export const TEST = 'test';

// TEST IDs
export const CARD = 'card';
export const MAIN_COMPONENT = 'main-component';
export const MAIN_TITLE = 'main-title';
export const MONTHLY_AMOUNT = 'monthly-amount';
export const SUMMARY_TEXT = 'summary-text';

// FUNCTIONS
export const getDefaultSavingGoal = (): SavingGoalState => {
  const today = getDate();
  const nextMonth = today.setMonth(today.getMonth() + 1, 1);
  return {
    goalAmount: INITIAL_GOAL_AMOUNT,
    reachDateYear: yearFormatter.format(nextMonth),
    reachDateMonth: monthFormatter.format(nextMonth),
    monthlyAmount: INITIAL_GOAL_AMOUNT,
    monthlyDeposits: 1,
  };
};
