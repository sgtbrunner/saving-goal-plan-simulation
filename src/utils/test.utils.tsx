import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import rootReducer from '../redux/root-reducer';
import { INITIAL_GOAL_AMOUNT } from './constants.utils';
import { getDate, monthFormatter, yearFormatter } from './functions.utils';
import { SavingGoalState } from './types.utils';

// TAG CONSTANTS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const IMG = 'img';
export const TEST = 'test';
export const AMOUNT = 'amount';

// TEST ID CONSTANTS
export const CARD = 'card';
export const MAIN_COMPONENT = 'main-component';
export const MAIN_TITLE = 'main-title';
export const MONTHLY_AMOUNT = 'monthly-amount';
export const SUMMARY_TEXT = 'summary-text';
export const CURRENCY_INPUT = 'currency-input';

// FUNCTIONS
export const getDefaultSavingGoal = (): SavingGoalState => {
  const today = getDate();
  const nextMonth = today.setMonth(today.getMonth() + 1, 1);
  return {
    goalAmount: INITIAL_GOAL_AMOUNT,
    reachDateYear: yearFormatter.format(nextMonth),
    reachDateMonth: monthFormatter.format(nextMonth),
    monthlyAmount: INITIAL_GOAL_AMOUNT,
    monthlyDeposits: 1,
  };
};

// RENDERER
type WrapperPropsType = {
  children: React.ReactNode | React.ReactNode[];
};

export const renderWithRedux = (
  component: React.ReactElement,
  store = createStore(rootReducer)
): RenderResult => {
  const ComponentWrapper = (props: WrapperPropsType) => (
    <Provider store={store}>{props.children}</Provider>
  );

  return render(component, {
    wrapper: ComponentWrapper as React.ComponentType,
  });
};
