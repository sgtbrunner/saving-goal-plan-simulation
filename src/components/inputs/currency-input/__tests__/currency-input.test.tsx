import {
  screen,
  RenderResult,
  fireEvent,
  waitFor,
} from '@testing-library/react';

import {
  AMOUNT,
  CURRENCY_INPUT,
  renderWithRedux,
} from '../../../../utils/test.utils';
import { CurrencyInputProps } from '../currency-input.types';
import CurrencyInput from '../currency-input.component';
import { ID } from '../../../../utils/constants.utils';

const defaultProps: CurrencyInputProps = {
  label: 'Total',
  updateAmount: jest.fn(),
};

const mountComponent = (newProps?: CurrencyInputProps): RenderResult =>
  renderWithRedux(<CurrencyInput {...defaultProps} {...newProps} />);

describe('CurrencyInput component', () => {
  it('should match snapshot', () => {
    const { container } = mountComponent();

    expect(container).toMatchSnapshot();
  });

  it(`should render input label with 'Total' text within the component`, () => {
    mountComponent();

    const label = screen.getByRole('textbox', { name: defaultProps.label });
    expect(label).toBeInTheDocument();
  });

  it('should render the input field with default value (25,000) within the component', () => {
    mountComponent();

    const inputField = screen.getByTestId(CURRENCY_INPUT);
    expect(inputField).toBeInTheDocument();
    expect(inputField).toHaveValue('25,000');
  });

  describe('user interaction events', () => {
    let inputField: HTMLElement;

    beforeEach(() => {
      mountComponent();

      inputField = screen.getByTestId(CURRENCY_INPUT);
      expect(inputField).toBeInTheDocument();
      expect(inputField).toHaveProperty(ID, AMOUNT);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should display the same value when the input is smaller than 1000', () => {
      fireEvent.change(inputField, { target: { value: '723' } });
      waitFor(() => {
        expect(inputField).toHaveValue('723');
        expect(defaultProps.updateAmount).toHaveBeenCalledTimes(1);
      });
    });

    it('should correctly format the value when the input is larger than 1000', () => {
      fireEvent.change(inputField, { target: { value: '2021' } });
      waitFor(() => {
        expect(inputField).toHaveValue('2,021');
        expect(defaultProps.updateAmount).toHaveBeenCalledTimes(1);
      });
    });

    it('should correctly format the value when the input is larger than 1000000', () => {
      fireEvent.change(inputField, { target: { value: '123456789' } });
      waitFor(() => {
        expect(inputField).toHaveValue('123,456,789');
        expect(defaultProps.updateAmount).toHaveBeenCalledTimes(1);
      });
    });

    it('should correctly format the value when the input has fraction digit', () => {
      fireEvent.change(inputField, { target: { value: '5006.48' } });
      waitFor(() => {
        expect(inputField).toHaveValue('5,006.48');
        expect(defaultProps.updateAmount).toHaveBeenCalledTimes(1);
      });
    });

    it('should correctly format the value to limit fracion digits when the input has many fraction digits', () => {
      fireEvent.change(inputField, { target: { value: '9547.123456' } });
      waitFor(() => {
        expect(inputField).toHaveValue('9,547.12');
        expect(defaultProps.updateAmount).toHaveBeenCalledTimes(1);
      });
    });
  });
});
