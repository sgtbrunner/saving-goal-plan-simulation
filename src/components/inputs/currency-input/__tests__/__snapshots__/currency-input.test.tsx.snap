// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput component should match snapshot 1`] = `
<div>
  <div>
    <label
      class="sc-bdnxRM eJleSl"
      for="amount"
    >
      Total
    </label>
    <div
      class="sc-gtsrHT eNnCka"
    >
      <div
        class="sc-dlnjwi hndFJ"
      >
        <span>
          $
        </span>
      </div>
      <input
        class="sc-hKFxyN kxEKCq"
        data-testid="currency-input"
        id="amount"
        inputmode="decimal"
        name="amount"
        type="text"
        value="25,000"
      />
    </div>
  </div>
</div>
`;
